File: project/app/src/main/java/org/owntracks/android/ui/map/MapViewModel.java
package org.owntracks.android.ui.map;

import android.location.Location;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.Bindable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.LocationSource;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.owntracks.android.data.repos.ContactsRepo;
import org.owntracks.android.injection.scopes.PerActivity;
import org.owntracks.android.model.FusedContact;
import org.owntracks.android.model.messages.MessageClear;
import org.owntracks.android.model.messages.MessageLocation;
import org.owntracks.android.services.LocationProcessor;
import org.owntracks.android.services.MessageProcessor;
import org.owntracks.android.support.Events;
import org.owntracks.android.support.SimpleIdlingResource;
import org.owntracks.android.ui.base.viewmodel.BaseViewModel;

import javax.inject.Inject;

import timber.log.Timber;

@PerActivity
public class MapViewModel extends BaseViewModel<MapMvvm.View> implements MapMvvm.ViewModel<MapMvvm.View>, LocationSource, GoogleMap.OnMapClickListener, GoogleMap.OnMarkerClickListener, GoogleMap.OnCameraMoveStartedListener {
    private final ContactsRepo contactsRepo;
    private final LocationProcessor locationProcessor;
    private FusedContact activeContact;
    private LocationSource.OnLocationChangedListener onLocationChangedListener;
    private MessageProcessor messageProcessor;
    private Location location;

    private static final int VIEW_FREE = 0;
    private static final int VIEW_CONTACT = 1;
    private static final int VIEW_DEVICE = 2;


    private static int mode = VIEW_DEVICE;
    private MutableLiveData<FusedContact> liveContact = new MutableLiveData<>();
    private MutableLiveData<Boolean> liveBottomSheetHidden = new MutableLiveData<>();
    private MutableLiveData<LatLng> liveCamera = new MutableLiveData<>();

    private final SimpleIdlingResource locationIdlingResource = new SimpleIdlingResource("locationIdlingResource", false);

    @Inject
    public MapViewModel(ContactsRepo contactsRepo, LocationProcessor locationRepo, MessageProcessor messageProcessor) {
        Timber.v("onCreate");
        this.contactsRepo = contactsRepo;
        this.messageProcessor = messageProcessor;
        this.locationProcessor = locationRepo;
    }

    @Override
    public void saveInstanceState(@NonNull Bundle outState) {
    }

    @Override
    public void restoreInstanceState(@NonNull Bundle savedInstanceState) {
    }

    @Override
    public LocationSource getMapLocationSource() {
        return this;
    }

    @Override
    public GoogleMap.OnMapClickListener getOnMapClickListener() {
        return this;
    }

    @Override
    public GoogleMap.OnMarkerClickListener getOnMarkerClickListener() {
        return this;
    }

    @Override
    public void onMapReady() {
        for (Object c : contactsRepo.getAll().getValue().values()) {
            getView().updateMarker((FusedContact) c);
        }

        if (mode == VIEW_CONTACT && activeContact != null)
            setViewModeContact(activeContact, true);
        else if (mode == VIEW_FREE) {
            setViewModeFree();
        } else {
            setViewModeDevice();
        }
    }

    @Override
    public LiveData<FusedContact> getContact() {
        return liveContact;
    }

    @Override
    public LiveData<Boolean> getBottomSheetHidden() {
        return liveBottomSheetHidden;
    }

    @Override
    public LiveData<LatLng> getCenter() {
        return liveCamera;
    }

    @Override
    public void sendLocation() {
        locationProcessor.publishLocationMessage(MessageLocation.REPORT_TYPE_USER);
    }


    private void setViewModeContact(@NonNull String contactId, boolean center) {
        FusedContact c = contactsRepo.getById(contactId);
        if (c != null)
            setViewModeContact(c, center);
        else
            Timber.e("contact not found %s, ", contactId);
    }

    private void setViewModeContact(@NonNull FusedContact c, boolean center) {
        mode = VIEW_CONTACT;
        Timber.v("contactId:%s, obj:%s ", c.getId(), activeContact);

        activeContact = c;

        liveContact.postValue(c);
        liveBottomSheetHidden.postValue(false);

        if (center)
            liveCamera.postValue(c.getLatLng());

    }

    private void setViewModeFree() {
        Timber.v("setting view mode: VIEW_FREE");
        mode = VIEW_FREE;
        clearActiveContact();
    }

    private void setViewModeDevice() {
        Timber.v("setting view mode: VIEW_DEVICE");

        mode = VIEW_DEVICE;
        clearActiveContact();
        if (hasLocation()) {
            liveCamera.postValue(getCurrentLocation());
        } else {
            Timber.e("no location available");
        }
    }

    @Override
    @Nullable
    public LatLng getCurrentLocation() {
        return location != null ? new LatLng(location.getLatitude(), location.getLongitude()) : null;
    }

    @Override
    @Bindable
    public FusedContact getActiveContact() {
        return activeContact;
    }


    @Override
    public void restore(@NonNull String contactId) {
        Timber.v("restoring contact id:%s", contactId);
        setViewModeContact(contactId, true);
    }

    @Override
    public boolean hasLocation() {
        return location != null;
    }


    private void clearActiveContact() {
        activeContact = null;
        liveContact.postValue(null);
        liveBottomSheetHidden.postValue(true);
    }

    @Override
    public void onBottomSheetClick() {
        getView().setBottomSheetExpanded();
    }

    @Override
    public void onMenuCenterDeviceClicked() {
        setViewModeDevice();
    }


    @Override
    public void onClearContactClicked() {
        MessageClear m = new MessageClear();
        if (activeContact != null) {
            m.setTopic(activeContact.getId());
            messageProcessor.queueMessageForSending(m);
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(Events.FusedContactAdded e) {
        onEvent(e.getContact());
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(Events.FusedContactRemoved c) {
        if (c.getContact() == activeContact) {
            clearActiveContact();
            setViewModeFree();
        }
        getView().removeMarker(c.getContact());
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(FusedContact c) {
        getView().updateMarker(c);
        if (c == activeContact) {
            liveContact.postValue(c);
            liveCamera.postValue(c.getLatLng());
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(Events.ModeChanged e) {
        getView().clearMarkers();
        clearActiveContact();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(Events.MonitoringChanged e) {
        getView().updateMonitoringModeMenu();
    }

    @Subscribe(threadMode = ThreadMode.MAIN, priority = 1, sticky = true)
    public void onEvent(@NonNull Location location) {
        this.location = location;
        getView().enableLocationMenus();
        locationIdlingResource.setIdleState(true);
        if (mode == VIEW_DEVICE) {
            liveCamera.postValue(getCurrentLocation());
        }
        if (onLocationChangedListener != null) {
            this.onLocationChangedListener.onLocationChanged(this.location);
        }
    }

    // Map Callback
    @Override
    public void activate(OnLocationChangedListener onLocationChangedListener) {
        Timber.v("location source activated");
        this.onLocationChangedListener = onLocationChangedListener;
        if (location != null)
            this.onLocationChangedListener.onLocationChanged(location);
    }

    // Map Callback
    @Override
    public void deactivate() {
        onLocationChangedListener = null;
    }

    // Map Callback
    @Override
    public void onMapClick(LatLng latLng) {
        setViewModeFree();
    }

    // Map Callback
    @Override
    public boolean onMarkerClick(Marker marker) {
        if (marker.getTag() != null) {
            setViewModeContact((String) marker.getTag(), false);
        }
        return true;
    }

    @Override
    public void onBottomSheetLongClick() {
        setViewModeContact(activeContact.getId(), true);
    }

    @Override
    public GoogleMap.OnCameraMoveStartedListener getOnMapCameraMoveStartedListener() {
        return this;
    }

    @Override
    public void onCameraMoveStarted(int reason) {
        if (reason == REASON_GESTURE) {
            setViewModeFree();
        }
    }

    public SimpleIdlingResource getLocationIdlingResource() {
        return locationIdlingResource;
    }
}

File: project/app/src/main/java/org/owntracks/android/ui/map/MapViewModel.kt
package org.owntracks.android.ui.map

import android.app.Application
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.location.Location
import androidx.annotation.MainThread
import androidx.databinding.Observable
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlin.math.asin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import org.owntracks.android.BR
import org.owntracks.android.data.repos.ContactsRepo
import org.owntracks.android.data.repos.ContactsRepoChange
import org.owntracks.android.data.repos.LocationRepo
import org.owntracks.android.data.waypoints.WaypointModel
import org.owntracks.android.data.waypoints.WaypointsRepo
import org.owntracks.android.geocoding.GeocoderProvider
import org.owntracks.android.location.LatLng
import org.owntracks.android.location.toLatLng
import org.owntracks.android.model.CommandAction
import org.owntracks.android.model.Contact
import org.owntracks.android.model.messages.MessageClear
import org.owntracks.android.model.messages.MessageCmd
import org.owntracks.android.model.messages.MessageLocation
import org.owntracks.android.preferences.Preferences
import org.owntracks.android.preferences.types.ConnectionMode
import org.owntracks.android.preferences.types.MonitoringMode
import org.owntracks.android.preferences.types.UnitsDisplay
import org.owntracks.android.services.LocationProcessor
import org.owntracks.android.services.MessageProcessor
import org.owntracks.android.support.Meters
import org.owntracks.android.support.RequirementsChecker
import timber.log.Timber

@HiltViewModel
class MapViewModel
@Inject
constructor(
    private val contactsRepo: ContactsRepo,
    private val locationProcessor: LocationProcessor,
    private val messageProcessor: MessageProcessor,
    private val geocoderProvider: GeocoderProvider,
    private val preferences: Preferences,
    private val locationRepo: LocationRepo,
    private val waypointsRepo: WaypointsRepo,
    application: Application,
    private val requirementsChecker: RequirementsChecker
) : AndroidViewModel(application) {
  // controls who the currently selected contact is
  private val mutableCurrentContact = MutableLiveData<Contact?>()
  val currentContact: LiveData<Contact?>
    get() = mutableCurrentContact

  // controls the state of the bottom sheet on the map
  private val mutableBottomSheetHidden = MutableLiveData<Boolean>()
  val bottomSheetHidden: LiveData<Boolean>
    get() = mutableBottomSheetHidden

  // Controls where the map should set the camera to
  private val mutableMapCenter = MutableLiveData<LatLng>()
  val mapCenter: LiveData<LatLng>
    get() = mutableMapCenter

  // Shows the current distance to the selected contact
  private val mutableContactDistance = MutableLiveData(Meters(0f))
  val contactDistance: LiveData<Meters>
    get() = mutableContactDistance

  // Shows the bearing to the selected contact
  private val mutableContactBearing = MutableLiveData(0f)
  val contactBearing: LiveData<Float>
    get() = mutableContactBearing

  // Shows the relative bearing from this device orientation to the contact
  private val mutableRelativeContactBearing = MutableLiveData(0f)
  val relativeContactBearing: LiveData<Float>
    get() = mutableRelativeContactBearing

  // Controls the current map layer style
  private val mutableMapLayerStyle = MutableLiveData(preferences.mapLayerStyle)
  val mapLayerStyle: LiveData<MapLayerStyle>
    get() = mutableMapLayerStyle

  // Controls the display units (metric/imperial)
  private val mutableUnitsDisplay = MutableLiveData(preferences.imperialUnitsDisplay)
  val unitsDisplay: LiveData<UnitsDisplay>
    get() = mutableUnitsDisplay

  // Controls the status of the MyLocation FAB on the map
  private val mutableMyLocationStatus = MutableLiveData(MyLocationStatus.DISABLED)
  val myLocationStatus: LiveData<MyLocationStatus>
    get() = mutableMyLocationStatus

  val currentLocation = LocationLiveData(application, viewModelScope)

  val waypointUpdatedEvent = waypointsRepo.repoChangedEvent

  suspend fun getAllWaypoints(): List<WaypointModel> {
    return waypointsRepo.getAll()
  }

  val allContacts = contactsRepo.all
  val contactUpdatedEvent: Flow<ContactsRepoChange> = contactsRepo.repoChangedEvent

  val scope: CoroutineScope
    get() = viewModelScope

  private val mutableCurrentMonitoringMode: MutableLiveData<MonitoringMode> by lazy {
    MutableLiveData(preferences.monitoring)
  }

  val currentMonitoringMode: LiveData<MonitoringMode>
    get() = mutableCurrentMonitoringMode

  private val currentConnectionMode: MutableLiveData<ConnectionMode> by lazy {
    MutableLiveData(preferences.mode)
  }

  val viewMode: ViewMode by locationRepo::viewMode

  val zoomLevel: Double?
    get() = locationRepo.mapViewWindowLocationAndZoom?.zoom

  /**
   * Sets the status of the "My Location" button based on whether the location permissions are
   * granted and what the current viewmode is. Whenever these things change, we should call this to
   * update
   */
  fun updateMyLocationStatus() {
    mutableMyLocationStatus.postValue(
        if (requirementsChecker.hasLocationPermissions() &&
            requirementsChecker.isLocationServiceEnabled()) {
          if (viewMode == ViewMode.Device) {
            MyLocationStatus.FOLLOWING
          } else {
            MyLocationStatus.AVAILABLE
          }
        } else {
          MyLocationStatus.DISABLED
        })
  }

  fun hasLocationPermission() = requirementsChecker.hasLocationPermissions()

  private val preferenceChangeListener =
      object : Preferences.OnPreferenceChangeListener {
        override fun onPreferenceChanged(properties: Set<String>) {
          if (properties.contains("monitoring")) {
            mutableCurrentMonitoringMode.postValue(preferences.monitoring)
          }
          if (properties.contains("mode")) {
            currentConnectionMode.postValue(preferences.mode)
            clearActiveContact()
          }
          if (properties.contains("imperialUnitsDisplay")) {
            mutableUnitsDisplay.postValue(preferences.imperialUnitsDisplay)
          }
        }
      }

  init {
    preferences.registerOnPreferenceChangedListener(preferenceChangeListener)
  }

  override fun onCleared() {
    super.onCleared()
    preferences.unregisterOnPreferenceChangedListener(preferenceChangeListener)
  }

  fun onMapReady() {
    when (viewMode) {
      is ViewMode.Contact -> {
        mutableCurrentContact.value?.run { setViewModeContact(this, true) }
      }
      is ViewMode.Free -> {
        setViewModeFree()
      }
      is ViewMode.Device -> {
        setViewModeDevice()
      }
    }
  }

  fun refreshGeocodeForContact(contact: Contact) {
    contact.geocodeLocation(geocoderProvider, viewModelScope)
  }

  fun sendLocation() {
    viewModelScope.launch {
      currentLocation.value?.run {
        Timber.d("Sending current location from user request: $this")
        locationProcessor.onLocationChanged(this, MessageLocation.ReportType.USER)
        locationProcessor.publishStatusMessage()
      }
    }
  }

  /** User has clicked on the MyLocation FAB, so we should set the viewMode to be device. */
  fun onMyLocationClicked() {
    setViewModeDevice()
  }

  private fun setViewModeContact(contactId: String, center: Boolean) {
    val c = contactsRepo.getById(contactId)
    if (c != null) {
      setViewModeContact(c, center)
    } else {
      Timber.e("contact not found %s, ", contactId)
    }
  }

  /**
   * We need a way of updating other bits in the viewmodel when the current contact's properties
   * change.
   *
   * @constructor Create empty Fused contact property changed callback
   */
  inner class ContactPropertyChangedCallback : Observable.OnPropertyChangedCallback() {
    override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
      sender?.run {
        if (this is Contact) {
          when (propertyId) {
            BR.latLng -> {
              updateActiveContactDistanceAndBearing(this)
            }
            BR.trackerId -> {
              mutableCurrentContact.postValue(this)
            }
          }
        }
      }
    }
  }

  private val contactPropertyChangedCallback = ContactPropertyChangedCallback()

  private fun setViewModeContact(contact: Contact, center: Boolean) {
    Timber.d("setting view mode: VIEW_CONTACT for $contact, center=$center")
    locationRepo.viewMode = ViewMode.Contact(center)
    mutableCurrentContact.value = contact
    contact.addOnPropertyChangedCallback(contactPropertyChangedCallback)
    mutableBottomSheetHidden.value = false
    refreshGeocodeForContact(contact)
    updateActiveContactDistanceAndBearing(contact)
    if (center && contact.latLng != null) mutableMapCenter.postValue(contact.latLng)
    updateMyLocationStatus()
  }

  private fun setViewModeFree() {
    Timber.d("setting view mode: VIEW_FREE")
    locationRepo.viewMode = ViewMode.Free
    clearActiveContact()
    updateMyLocationStatus()
  }

  private fun setViewModeDevice() {
    Timber.d("setting view mode: VIEW_DEVICE")
    locationRepo.viewMode = ViewMode.Device
    clearActiveContact()
    currentLocation.value?.apply { mutableMapCenter.postValue(this.toLatLng()) }
        ?: run { Timber.w("no location available") }
    updateMyLocationStatus()
  }

  @MainThread
  fun setLiveContact(contactId: String?) {
    contactId?.let {
      locationRepo.viewMode = ViewMode.Contact(true)
      contactsRepo.getById(it)?.run(mutableCurrentContact::setValue)
    }
  }

  private fun clearActiveContact() {
    mutableCurrentContact.value?.removeOnPropertyChangedCallback(contactPropertyChangedCallback)
    mutableCurrentContact.postValue(null)
    mutableBottomSheetHidden.postValue(true)
  }

  fun onClearContactClicked() {
    mutableCurrentContact.value?.also {
      messageProcessor.queueMessageForSending(MessageClear().apply { topic = it.id })
      viewModelScope.launch { contactsRepo.remove(it.id) }
    }
    clearActiveContact()
  }

  private val mutableLocationRequestContactCommandFlow =
      MutableSharedFlow<Contact>(extraBufferCapacity = 1)

  val locationRequestContactCommandFlow: Flow<Contact> = mutableLocationRequestContactCommandFlow

  fun sendLocationRequestToCurrentContact() {
    mutableCurrentContact.value?.also {
      messageProcessor.queueMessageForSending(
          MessageCmd().apply {
            topic = it.id + preferences.commandTopicSuffix
            action = CommandAction.REPORT_LOCATION
          })
      mutableLocationRequestContactCommandFlow.tryEmit(it)
    }
  }

  private fun updateActiveContactDistanceAndBearing(contact: Contact) {
    currentLocation.value?.run { updateActiveContactDistanceAndBearing(this, contact) }
  }

  fun updateActiveContactDistanceAndBearing(currentLocation: Location) {
    mutableCurrentContact.value?.run {
      updateActiveContactDistanceAndBearing(currentLocation, this)
    }
  }

  private fun updateActiveContactDistanceAndBearing(currentLocation: Location, contact: Contact) {
    contact.latLng?.run {
      val distanceBetween = FloatArray(2)
      Location.distanceBetween(
          currentLocation.latitude,
          currentLocation.longitude,
          latitude.value,
          longitude.value,
          distanceBetween)
      mutableContactDistance.postValue(Meters(distanceBetween[0]))
      mutableContactBearing.postValue(distanceBetween[1])
      mutableRelativeContactBearing.postValue(distanceBetween[1])
    }
  }

  fun onMapClick() {
    setViewModeFree()
  }

  /**
   * User has clicked on a marker on the map. Should set the view mode to contact, but not center
   * the map
   *
   * @param id identifier of the marker that was clicked on
   */
  fun onMarkerClick(id: String) {
    setViewModeContact(id, false)
  }

  /**
   * User has long clicked on the bottom contact sheet. This should trigger centering the map on the
   * currently active contact
   */
  fun onBottomSheetLongClick() {
    mutableCurrentContact.value?.run { setViewModeContact(id, true) }
  }

  /** Start requesting location updates for the blue dot */
  fun requestLocationUpdatesForBlueDot() {
    if (requirementsChecker.hasLocationPermissions()) {
      @Suppress("MissingPermission") // We've already checked for permissions
      viewModelScope.launch { currentLocation.requestLocationUpdates() }
    }
  }

  /**
   * Moves the blue dot on the map to the given location
   *
   * @param latLng location to move the blue dot to
   */
  fun setCurrentBlueDotLocation(latLng: LatLng) {
    locationRepo.currentBlueDotOnMapLocation = latLng
  }

  /**
   * Updates the stored mapView to the given location zoom and rotation. Triggered when the map is
   * moved from the [MapFragment]
   *
   * @param mapLocationZoomLevelAndRotation the current map location, zoom and rotation values
   */
  fun setMapLocationFromMapMoveEvent(
      mapLocationZoomLevelAndRotation: MapLocationZoomLevelAndRotation
  ) {
    locationRepo.mapViewWindowLocationAndZoom = mapLocationZoomLevelAndRotation
  }

  fun initMapStartingLocation(): MapLocationZoomLevelAndRotation =
      if (viewMode == ViewMode.Contact(true) && currentContact.value?.latLng != null) {
        MapLocationZoomLevelAndRotation(
            currentContact.value!!.latLng!!,
            locationRepo.mapViewWindowLocationAndZoom?.zoom ?: STARTING_ZOOM)
      } else {
        locationRepo.mapViewWindowLocationAndZoom
            ?: locationRepo.currentBlueDotOnMapLocation?.let {
              MapLocationZoomLevelAndRotation(it, STARTING_ZOOM)
            }
            ?: locationRepo.currentPublishedLocation.value?.let {
              MapLocationZoomLevelAndRotation(it.toLatLng(), STARTING_ZOOM)
            }
            ?: MapLocationZoomLevelAndRotation(
                LatLng(STARTING_LATITUDE, STARTING_LONGITUDE), STARTING_ZOOM)
      }

  val orientationSensorEventListener =
      object : SensorEventListener {
        override fun onSensorChanged(maybeEvent: SensorEvent?) {
          maybeEvent?.let { event ->
            currentContact.value?.latLng?.let { contactLatLng ->
              currentLocation.value?.let { currentLocation ->
                // Orientation is angle around the Z axis
                val azimuth = (180 / Math.PI) * 2 * asin(event.values[2])
                val distanceBetween = FloatArray(2)
                Location.distanceBetween(
                    currentLocation.latitude,
                    currentLocation.longitude,
                    contactLatLng.latitude.value,
                    contactLatLng.longitude.value,
                    distanceBetween)
                mutableRelativeContactBearing.postValue(distanceBetween[1] + azimuth.toFloat())
              }
            }
          }
        }

        override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
          // noop
        }
      }

  fun setMonitoringMode(mode: MonitoringMode) {
    preferences.monitoring = mode
  }

  fun setMapLayerStyle(mapLayerStyle: MapLayerStyle) {
    preferences.mapLayerStyle = mapLayerStyle
    mutableMapLayerStyle.postValue(mapLayerStyle)
  }

  companion object {
    // Paris
    private const val STARTING_LATITUDE = 48.856826
    private const val STARTING_LONGITUDE = 2.292713
    private const val STARTING_ZOOM = 15.0
  }

  sealed class ViewMode {
    object Free : ViewMode()

    object Device : ViewMode()

    data class Contact(val follow: Boolean) : ViewMode()
  }
}
